packages/angular/src/jsonforms-root.component.ts

```diff
@@ -0,0 +1,116 @@
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  THE SOFTWARE.
+*/
+import {
+    Component,
+    EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges
+} from '@angular/core';
+import { Actions, JsonFormsRendererRegistryEntry, JsonSchema, UISchemaElement, UISchemaTester } from '@jsonforms/core';
+import { JsonFormsAngularService } from './jsonforms.service';
+@Component({
+    selector: 'jsonforms',
+    template: '<jsonforms-outlet></jsonforms-outlet>',
+    providers: [JsonFormsAngularService]
+})
+export class JsonForms implements OnChanges, OnInit {
+
+    @Input() uischema: UISchemaElement;
+    @Input() schema: JsonSchema;
+    @Input() data: any;
+    @Input() renderers: JsonFormsRendererRegistryEntry[];
+    @Input() uischemas: { tester: UISchemaTester; uischema: UISchemaElement; }[];
+    @Output() dataChange = new EventEmitter<any>();
+    @Input() locale: string;
+    @Input() readonly: boolean;
+
+    private initialized = false;
+
+    constructor(private jsonformsService: JsonFormsAngularService) {
+    }
+
+    ngOnInit(): void {
+        this.jsonformsService.init({
+            core: {
+                data: this.data,
+                uischema: this.uischema,
+                schema: this.schema
+            },
+            uischemas: this.uischemas,
+            i18n: { locale: this.locale, localizedSchemas: undefined, localizedUISchemas: undefined },
+            renderers: this.renderers
+        });
+        this.jsonformsService.$state.subscribe(state => {
+            const data = state?.jsonforms?.core?.data;
+            if (data) {
+                this.dataChange.emit(data);
+            }
+        });
+        this.initialized = true;
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        if (!this.initialized) {
+            return;
+        }
+        const newData = changes.data;
+        const newSchema = changes.schema;
+        const newUiSchema = changes.uischema;
+        const newRenderers = changes.renderers;
+        const newUischemas = changes.uischemas;
+        const newLocale = changes.locale;
+        const newReadonly = changes.readonly;
+
+        if (newData && newSchema && newUiSchema) {
+            this.jsonformsService.updateCore(
+                Actions.updateCore(newData.currentValue, newSchema.currentValue, newUiSchema.currentValue)
+            );
+        } else {
+            if (newData && !newData.isFirstChange()) {
+                this.jsonformsService.setData(newData.currentValue);
+            }
+            if (newSchema && !newSchema.isFirstChange()) {
+                this.jsonformsService.setSchema(newSchema.currentValue);
+            }
+            if (newUiSchema && !newUiSchema.isFirstChange()) {
+                this.jsonformsService.setUiSchema(newUiSchema.currentValue);
+            }
+        }
```

Comment:
We don't optimize now for cases when only two of those attributes change. It could see this happen in cases where the user never specifies a ui schema but exchanges data and schema.

This is more of an edge case and only leads to a small performance hit instead of a misbehavior so we could leave it like this for now.

Reply:
Ah actually I missed a problem here: In the first case when all three props change and the new `schema` or `uiSchema` is undefined they will not be newly generated and the app will break.

If you don't want to go with the approach I originally suggested because we're then generating schema and or ui schema here then we actually need a new method in the forms service which we call which does it for us. Sadly the most obvious name `updateCore` is already used for the dispatching itself ;)